pkg/common/log/reopen_test.go

```diff
@@ -0,0 +1,115 @@
+package log
+
+import (
+	"context"
+	"errors"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/spiffe/spire/test/spiretest"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestReopenOnSignal(t *testing.T) {
+	const (
+		_tempDir         = "spirelogrotatetest"
+		_testLogFileName = "test.log"
+		_rotatedSuffix   = "rotated"
+		_firstMsg        = "a message"
+		_secondMsg       = "another message"
+	)
+	dir := spiretest.TempDir(t)
+
+	logFileName := filepath.Join(dir, _testLogFileName)
+	rotatedLogFileName := logFileName + "." + _rotatedSuffix
+	rwc, err := NewReopenableFile(logFileName)
+	require.NoError(t, err)
+
+	fsInfo, err := rwc.f.Stat()
+	require.NoError(t, err)
+	assert.Equal(t, int64(0), fsInfo.Size(), "%s should be empty", fsInfo.Name())
+
+	logger, err := NewLogger(WithReopenableOutputFile(rwc))
+	require.NoError(t, err)
+
+	logger.Warning(_firstMsg)
+
+	fsInfo, err = rwc.f.Stat()
+	require.NoError(t, err)
+	initialLogSize := fsInfo.Size()
+	initialLogModTime := fsInfo.ModTime()
+	assert.NotEqual(t, int64(0), fsInfo.Size(), "%s should not be empty", fsInfo.Name())
+
+	ctx, cancel := context.WithCancel(context.Background())
+	signalCh := make(chan os.Signal, 1)
+
+	renamedCh := make(chan struct{})
+	go func() {
+		// emulate logrotate
+		err = os.Rename(logFileName, rotatedLogFileName)
+		require.NoError(t, err)
+		signalCh <- _reopenSignal
+		// explicitly cancel so test continues
+		cancel()
```

Comment:
Since signalCh is buffered, I'm not sure we have a guarantee about which channel `reopenOnSignal` will select on (i.e. the signal channel or the context). As such, this seems racy or prone to break with underlying changes to the scheduler.

I wonder if a more stable way of ensuring that the reopen is processed completely is to wrap the `rwc` in a `Reopener` that calls `Reopen` on the wrapped `Reopener` and then cancels the context.

Reply:
I was struggling with it a bit here. I did high count under race detector and this appears stable as written.

> prone to break with underlying changes to the scheduler

True.

I will try your wrapper suggestion.

Reply:
That works a charm.